libs/common-types/src/lib.rs

```diff
@@ -130,47 +197,166 @@ where
+				PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.remove(id, delta),
+			},
+			Role::PermissionedCurrencyRole(permissioned_currency_role) => {
+				match permissioned_currency_role {
+					PermissionedCurrencyRole::Holder(delta) => {
+						self.permissioned_asset_holder.remove(delta)
+					}
+					PermissionedCurrencyRole::Manager => {
+						Ok(self.admin.remove(AdminRoles::PERMISSIONED_ASSET_MANAGER))
+					}
+					PermissionedCurrencyRole::Issuer => {
+						Ok(self.admin.remove(AdminRoles::PERMISSIONED_ASSET_ISSUER))
+					}
+				}
+			}
 		}
 	}
 
 	fn add(&mut self, property: Self::Property) -> Result<(), ()> {
 		match property {
-			PoolRole::Borrower => Ok(self.admin.insert(AdminRoles::BORROWER)),
-			PoolRole::LiquidityAdmin => Ok(self.admin.insert(AdminRoles::LIQUIDITY_ADMIN)),
-			PoolRole::PoolAdmin => Ok(self.admin.insert(AdminRoles::POOL_ADMIN)),
-			PoolRole::PricingAdmin => Ok(self.admin.insert(AdminRoles::PRICING_ADMIN)),
-			PoolRole::MemberListAdmin => Ok(self.admin.insert(AdminRoles::MEMBER_LIST_ADMIN)),
-			PoolRole::RiskAdmin => Ok(self.admin.insert(AdminRoles::RISK_ADMIN)),
-			PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.insert(id, delta),
+			Role::PoolRole(pool_role) => match pool_role {
+				PoolRole::Borrower => Ok(self.admin.insert(AdminRoles::BORROWER)),
+				PoolRole::LiquidityAdmin => Ok(self.admin.insert(AdminRoles::LIQUIDITY_ADMIN)),
+				PoolRole::PoolAdmin => Ok(self.admin.insert(AdminRoles::POOL_ADMIN)),
+				PoolRole::PricingAdmin => Ok(self.admin.insert(AdminRoles::PRICING_ADMIN)),
+				PoolRole::MemberListAdmin => Ok(self.admin.insert(AdminRoles::MEMBER_LIST_ADMIN)),
+				PoolRole::RiskAdmin => Ok(self.admin.insert(AdminRoles::RISK_ADMIN)),
+				PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.insert(id, delta),
+			},
+			Role::PermissionedCurrencyRole(permissioned_currency_role) => {
+				match permissioned_currency_role {
+					PermissionedCurrencyRole::Holder(delta) => {
+						self.permissioned_asset_holder.insert(delta)
+					}
+					PermissionedCurrencyRole::Manager => {
+						Ok(self.admin.insert(AdminRoles::PERMISSIONED_ASSET_MANAGER))
+					}
+					PermissionedCurrencyRole::Issuer => {
+						Ok(self.admin.insert(AdminRoles::PERMISSIONED_ASSET_ISSUER))
+					}
+				}
+			}
+		}
+	}
+}
+
+impl<Now, MinDelay, Moment> PermissionedCurrencyHolders<Now, MinDelay, Moment>
+where
+	Now: UnixTime,
+	MinDelay: Get<Moment>,
+	Moment: From<u64> + PartialEq + PartialOrd + Saturating + Ord + Copy,
+{
+	pub fn empty() -> Self {
+		Self::default()
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.info.is_none()
+	}
+
+	fn validity(&self, delta: Moment) -> Result<Moment, ()> {
+		let now: Moment = Now::now().as_secs().into();
+		let min_validity = now.saturating_add(MinDelay::get());
+		let req_validity = now.saturating_add(delta);
+
+		if req_validity < min_validity {
+			Err(())
+		} else {
+			Ok(req_validity)
+		}
```

Comment:
Doesn't matter but in many other places we and Substrate overall opts for this style instead:

```suggestion
		if req_validity < min_validity {
			return Err(());
		}
		
		Ok(req_validity)
```